fs/io_zenfs.cc

```diff
@@ -371,7 +379,13 @@ IOStatus ZoneFile::PositionedRead(uint64_t offset, size_t n, Slice* result,
       break;
     }
 
-    pread_sz = (size_t)r;
+    /* Verify and update the the bytes read count (if read size was incremented,
+     * for alignment purposes).
+     */
+    if ((size_t)r <= pread_sz - bytes_to_align)
+      pread_sz = (size_t)r;
+    else
+      pread_sz -= bytes_to_align;
```

Comment:
This won't work if only part of the alignment area has been read. This *should* not happen for direct reads but let's make sure we don't fail to calculate the right result in that case anyway.

Keep pread_sz indicating the valid data bytes and introduce a pread_sz_aligned varible in stead and use that for the direct read. then it is enough to do this:

if (r < pread_sz  ) pread_sz = r;
(when r >= pread_sz we don't need to do anything)

Reply:
Just to understand the case we are discussing here, if the initial pread_sz is 6K, then we align pread_sz to 8K (bytes_to_align=2K), then pread() returns 7K as the bytes read.
I considered this case where only part of alignment area is read and here I return, the bytes requested initially, (i.e. 6K (8K - 2k) in this example), so that it is like we returned exactly what was initially requested and whatever extra bytes were read due to alignment adjustment is just ignored. 

Even with 
r = pread(f_direct, ptr, pread_sz_aligned, r_off);
if (r < pread_sz ) pread_sz = r;

If r was 7K, then pread_sz will be 6K. So in both cases, 6K  will be the bytes read(for accounting purposes).
So I think the current implementation behaves the same way as suggested changes.

Or, are you suggesting we do;
if (r < pread_sz_aligned) pread_sz = r;
To return 7K when 6K was the initial request.

Reply:
Ah, yes, you're right.